### PR TITLE
Added Kallisto WDL

### PR DESCRIPTION
**Pull request by ambrosejcarr**
This PR adds: 

1. Kallisto task WDL
2. Kallisto mkref workflow
3. Kallisto testing notebook
4. Kallisto testing WDL

The main deliverable here is the task WDL, the rest are accessories involved in validating it's functionality.

### CODE REVIEW

File: pipelines/tasks/Kallisto.wdl [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  # estimate disk requirements
+  Int disks = 10 + ceil(size(r1, "G") * 4 + size(index, "G") * 2)
+
+  command {
+    kallisto quant \
+      --index "${index}" \
+      --output-dir . \
+      --bootstrap-samples 100 \
+      --single \
+      --fragment-length 250 \
```
Comment (jishuxu) [inline pipelines/tasks/Kallisto.wdl]: what's the fragment  estimation from? from 10X report or just arbitrary number?

File: pipelines/tasks/Kallisto.wdl [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+task QuantPairedEnd {
+  File r1  # forward read
+  File index  # kallisto index, output of Kallisto.Mkref task
+  File r2  # reverse read
+
+  # estimate disk requirements
+  Int disks = 10 + ceil((size(r1, "G") + size(r2, "G")) * 4 + size(index, "G") * 2)
+
+  command {
+    kallisto quant \
```
Comment (jishuxu) [inline pipelines/tasks/Kallisto.wdl]: Can you add '--Pseudobam'  in cmd and output pseudoalignment bam file. cmd is something like 
kallisto quant -i index -o out --pseudobam r1.fastq r2.fastq | samtools view -Sb - > out.bam

File: pipelines/tasks/Kallisto.wdl [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+      --index "${index}" \
+      --output-dir . \
+      --bootstrap-samples 100 \
+      "${r1}" \
+      "${r2}"
+  }
+
+  runtime {
+    docker: "humancellatlas/kallisto:0.43.1"
+    cpu: 8
```
Comment (jishuxu) [inline pipelines/tasks/Kallisto.wdl]: Should you add '--threads=8 '  in cmd? The default thread num is 1.

File: pipelines/tasks/Kallisto.wdl [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+  output {
+    File abundance_h5 = "abundance.h5"
+    File abundance_tsv = "abundance.tsv"
+    File log = "run_info.json"
+  }
+}
+
+
+task PseudoSingleEnd {
```
Comment (jishuxu) [inline pipelines/tasks/Kallisto.wdl]: what's the difference  between 'pseudoSingleEnd' and 'QuantSingleEnd'?

File: pipelines/tasks/Kallisto.wdl [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+  output {
+    File pseudo_ec = "pseudoalignments.ec"
+    File pseudo_tsv = "pseudoalignments.tsv"
+    File run_log = "run_info.json"
+  }
+}
+
+
+task PseudoSingleEndUMI {
```
Comment (jishuxu) [inline pipelines/tasks/Kallisto.wdl]: what's the difference  between 'PseudoSingleEndUMI' and 'PseudoSingleEnd' ?

File: pipelines/tasks/Kallisto.wdl [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  File index  # kallisto index, output of Kallisto.Mkref task
+
+  # estimate disk requirements
+  Int disks = 10 + ceil(size(r1, "G") * 4 + size(index, "G") * 2)
+
+  command {
+    kallisto quant \
+      --index "${index}" \
+      --output-dir . \
+      --bootstrap-samples 100 \
```
Comment (jishuxu) [inline pipelines/tasks/Kallisto.wdl]: I would add '--seed' to have fixed random number. it is just easier from testing point of view.

**Comment by jishuxu** [comment]
Overall it is great. the testing notebook is informative. 
Is it possible to check Kallisto quantification results, either by printing out several lines of count matrix or visualizing the count matrix.  I just want to make sure the Kallisto actually produces the results we expect.

File: pipelines/tasks/Kallisto.wdl (hunk 2 above)
Comment (ambrosejcarr) [reply pipelines/tasks/Kallisto.wdl]: Finding alllll them undocumented requirements!

```
[quant] fragment length distribution will be estimated from the data
Error: pseudobam is not compatible with running on many threads.
```

File: pipelines/tasks/Kallisto.wdl (hunk 1 above)
Comment (ambrosejcarr) [reply pipelines/tasks/Kallisto.wdl]: arbitrary; updated now to be 10x numbers, as variables. 👍

File: pipelines/tasks/Kallisto.wdl (hunk 3 above)
Comment (ambrosejcarr) [reply pipelines/tasks/Kallisto.wdl]: Added where possible; pseudobam does not support multiple threads, as discovered above.

File: pipelines/tasks/Kallisto.wdl (hunk 4 above)
Comment (ambrosejcarr) [reply pipelines/tasks/Kallisto.wdl]: Pseudo only runs pseudoalignment. Quant runs both Pseudoalignment and quantification. See: http://pachterlab.github.io/kallisto/manual.html

File: pipelines/tasks/Kallisto.wdl (hunk 5 above)
Comment (ambrosejcarr) [reply pipelines/tasks/Kallisto.wdl]: UMI counts number of **UMIs** in each compatibility group, the other counts the number of **reads**.

File: pipelines/tasks/Kallisto.wdl (hunk 6 above)
Comment (ambrosejcarr) [reply pipelines/tasks/Kallisto.wdl]: Default of 42 should produce a fixed seed, no? Is there a point in setting it to something else that I'm missing?

**Comment by ambrosejcarr** [comment]
Added a section to the notebook to visualize the results of each function (except index, whose functionality is implied by the successful exit of dependent methods). 

Added docstrings to each wdl method to make it clearer what they're designed to do.

File: pipelines/tasks/Kallisto.wdl (hunk 6 above)
Comment (jishuxu) [reply pipelines/tasks/Kallisto.wdl]: copy paste from manual "--seed=INT                Seed for the bootstrap sampling (default: 42)"
http://pachterlab.github.io/kallisto/manual.html

File: pipelines/tasks/Kallisto.wdl (hunk 2 above)
Comment (jishuxu) [reply pipelines/tasks/Kallisto.wdl]: they should do better job on their documentation!!!

**Comment by jishuxu** [comment]
The updated notebook is exactly what I asked for. Thanks!

**Comment by jishuxu** [comment]
PR looks good fine with me and it is okay to merge

**Pull request merged**